File: src/spec/resources/testprojects/nodetest/src/auth/oauth/oauth2-password-authenticator.ts
import { BaseAuthenticator } from '../base-authenticator.js';
import { OAuth2TokenManager } from './oauth2-token-manager.js';

/**
 * Authenticator for the OAuth2 Resource Owner Password flow.
 */
export class OAuth2PasswordAuthenticator extends BaseAuthenticator {
  private readonly host: string;
  private readonly clientId: string;
  private readonly clientSecret: string;
  private readonly tokenUrl: string;
  private readonly username: string;
  private readonly password: string;
  private readonly scopes: readonly string[];
  private readonly tokenManager: OAuth2TokenManager;

  constructor(
    host: string,
    clientId: string,
    clientSecret: string,
    tokenUrl: string,
    username: string,
    password: string,
    scopes: string[]
  ) {
    super();
    this.host = host;
    this.clientId = clientId;
    this.clientSecret = clientSecret;
    this.tokenUrl = tokenUrl;
    this.username = username;
    this.password = password;
    this.scopes = Object.freeze([...scopes]);
    this.tokenManager = new OAuth2TokenManager();
  }

  getHost(): string {
    return this.host;
  }

  getAuthHeaders(): Record<string, string> {
    throw new Error('Use getAuthHeadersAsync() instead');
  }

  async getAuthHeadersAsync(): Promise<Record<string, string>> {
    const params: Record<string, string> = {
      grant_type: 'password',
      client_id: this.clientId,
      client_secret: this.clientSecret,
      username: this.username,
      password: this.password
    };
    if (this.scopes.length > 0) {
      params.scope = this.scopes.join(' ');
    }
    const token = await this.tokenManager.getAccessToken(this.tokenUrl, params);
    return { Authorization: `Bearer ${token}` };
  }
}

File: src/spec/resources/testprojects/csharptest/src/PetstoreClient/Auth/OAuth/OAuth2PasswordAuthenticator.cs
#pragma warning disable IDE0290 // Use primary constructor

namespace PetstoreClient.Auth.OAuth;

/// <summary>
/// Authenticator for the OAuth2 Resource Owner Password Credentials flow.
/// </summary>
public sealed class OAuth2PasswordAuthenticator : BaseAuthenticator
{
    private readonly string _host;
    private readonly string _clientId;
    private readonly string _clientSecret;
    private readonly Uri _tokenUrl;
    private readonly string _username;
    private readonly string _password;
    private readonly string[] _scopes;
    private readonly OAuth2TokenManager _tokenManager = new();

    public OAuth2PasswordAuthenticator(
        string host,
        string clientId,
        string clientSecret,
        Uri tokenUrl,
        string username,
        string password,
        string[] scopes
    )
    {
        _host = host;
        _clientId = clientId;
        _clientSecret = clientSecret;
        _tokenUrl = tokenUrl;
        _username = username;
        _password = password;
        _scopes = [.. scopes];
    }

    /// <inheritdoc/>
    public override string GetHost()
    {
        return _host;
    }

    /// <inheritdoc/>
    public override Dictionary<string, string> GetAuthHeaders()
    {
        Dictionary<string, string> parameters = new()
        {
            ["grant_type"] = "password",
            ["client_id"] = _clientId,
            ["client_secret"] = _clientSecret,
            ["username"] = _username,
            ["password"] = _password,
        };
        if (_scopes.Length > 0)
        {
            parameters["scope"] = string.Join(" ", _scopes);
        }

        string token = _tokenManager
            .GetAccessTokenAsync(_tokenUrl, parameters)
            .GetAwaiter()
            .GetResult();
        return new() { ["Authorization"] = "Bearer " + token };
    }
}
